src/os/bluestore/BlueFS.cc

```diff
@@ -1401,14 +1415,18 @@ int BlueFS::_flush_range(FileWriter *h, uint64_t offset, uint64_t length)
     h->file->fnode.mtime = ceph_clock_now(NULL);
     log_t.op_file_update(h->file->fnode);
     if (h->file->dirty_seq == 0) {
-      dirty_files.push_back(*h->file);
       dout(20) << __func__ << " dirty_seq = " << log_seq + 1
 	       << " (was clean)" << dendl;
     } else {
+      // already in dirty list, erase first
+      assert(dirty_files.count(h->file->dirty_seq));
+      auto it = dirty_files[h->file->dirty_seq].iterator_to(*h->file);
+      dirty_files[h->file->dirty_seq].erase(it);
```

Comment:
this could check for dirty_seq == log_seq+1 so that we don't remove+add in the case that the log_seq is the same.